multiphenicsx/fem/petsc.py

```diff
@@ -1573,34 +1574,67 @@ def set_bc(  # type: ignore[no-any-unimported]
     b: petsc4py.PETSc.Vec, bcs: list[dolfinx.fem.DirichletBC] = [],
     x0: typing.Optional[petsc4py.PETSc.Vec] = None,
     alpha: float = 1.0,
-    restriction: typing.Optional[mcpp.fem.DofMapRestriction] = None
+    restriction: typing.Optional[mcpp.fem.DofMapRestriction] = None,
+    restriction_x0: typing.Optional[mcpp.fem.DofMapRestriction] = None,
 ) -> None:
-    """Apply the function :func:`dolfinx.fem.set_bc` to a PETSc Vector."""
+    """Apply the function :func:`dolfinx.fem.set_bc` to a PETSc Vector:
+
+    For each Dirichet condition in bcs, it modifies b such that:
+
+    .. math::
+
+        b \\leftarrow  \\alpha (u_{bc} - x_0)
+
+    where :math:`u_{bc}` contains the Dirichlet values.
+
+    Parameters
+    ----------
+        b
+            PETSc Vector
+        bcs
+            Dirichlet conditions
+        x0
+            PETSc Vector subtracted to the Dirichlet values.
+        alpha
+            Scaling factor
+        restriction, restriction_x0 
+            Dofmap restrictions for b and x0. If not provided, the unrestricted vectors will be used for each.
+    """
+    @contextmanager
+    def x0_wrapper():
```

Comment:
Is this really needed? Isn't `VecSubVectorReadWrapper` on its own smart enough to handle the `None` cases? I thought I had all of the combinations of `None` vs not `None` handled there.

Reply:
You are right. I hyperfocused on how to get a dofmap if x0 was None.